src/java/org/apache/cassandra/io/sstable/SSTable.java

```diff
@@ -335,6 +338,16 @@ public String toString()
         return components;
     }
 
+    /**
+     * Rewrite TOC components by deleting existing TOC file and append new components
+     */
+    private static void rewriteTOC(Descriptor descriptor, Collection<Component> components)
+    {
+        File tocFile = descriptor.fileFor(Component.TOC);
+        tocFile.delete();
```

Comment:
I'm having a bit of trouble following the locking semantics of `rewriteTOC` and `appendTOC`. Can you help me understand how it's impossible to call both in parallel and lose updates if we race?

Reply:
Interesting question. The answer is I'm not sure if it is impossible. I've been through the paths for each of these and while `rewriteTOC` is only called from a `synchronized` method `appendTOC` isn't.

This code has been running in 6.8 production for a while so I'm assuming there isn't an issue with it but I will dig into it further to try and determine whether we need further locking round this file.

Reply:
The problem is that now you're doing a delete on the TOC file which wasn't happening before and so there's a new race :cry:

Reply:
@mfleming I've been over this and there is a potential race during a snapshot and an index getting removed. It's a very small window but it does need closing. If a manual snapshot and a index drop were done at the same time the index drop could potentially delete the TOC before the snapshot had read it, in which case the snapshot would start writing the TOC at the same time as the index drop.

The only other place `componentsFor` is called (outside of testing) is in the creation of SSTables so there can't be a conflict there.

I have added a `synchronized(descriptor)` inside both the `rewriteTOC` and appendTOC` methods to avoid this race.

Reply:
Sounds good